dlr_vieten/explorer/templates/dlr_vieten_explorer_tolerance_factors.html

```diff
@@ -25,17 +26,202 @@
+                line.append(element);
+                var wrapper = $('<div>'); wrapper.addClass('col-md-3');
+                var element = $('<div>'); element.addClass("input-group");
+                var subelement = $('<span>');
+                subelement.addClass('input-group-addon').html('Stoichiometry');
+                element.append(subelement);
+                var subelement = $('<input>');
+                subelement.addClass('form-control').prop('type', 'text')
+                            .prop('placeholder', '0 ... 1');
+                element.append(subelement);
+                wrapper.append(element);
+                line.append(wrapper);
+                var data = {};
+                for (i = 0; i < table['rows'].length; i++) {
+                    var row = table['rows'][i];
+                    if (row['type'] == 'transition metal' && row['species'] == this.value) {
+                        var c = row['charge'];
+                        if (!(c in data)) { data[c] = {}; }
+                        data[c][row['HS/LS']] = row['rᵢₒₙ'];
+                    }
+                }
+                var charges = ["3+", "4+"];
+                for (i = 0; i < charges.length; i++) {
+                    c = charges[i];
+                    if (c in data) {
+                        var wrapper = $('<div>'); wrapper.addClass('col-md-2');
+                        var element = $('<div>'); element.addClass("input-group");
+                        var subelement = $('<span>');
+                        subelement.addClass('input-group-addon').html(c);
+                        element.append(subelement);
+                        var subelement = $('<input>');
+                        var first_key = Object.keys(data[c])[0];
+                        subelement.addClass('form-control').prop('type', 'text')
+                            .prop('placeholder', data[c][first_key]);
+                        element.append(subelement);
+                        var subelement = $('<span>');
+                        subelement.addClass('input-group-addon').html(first_key);
+                        element.append(subelement);
+                        wrapper.append(element);
+                        line.append(wrapper);
+                    }
+                };
+                $('#listTM').append(line);
+            });
+            $('#go1').on('click', function (e) {
+                var t = parseFloat($('#t').attr('placeholder'));
+                var r = parseFloat($('#r').attr('placeholder'));
+                var A_avg_Mg = 0; var A_avg_Ca = 0; var A_avg_Sr = 0; var A_avg_Ba = 0; var rO;
+                for (i = 0; i < table['rows'].length; i++) {
+                    var row = table['rows'][i];
+                    if (row['type'] == 'alkali earth') {
+                        var r_ion = parseFloat(row['rᵢₒₙ'].split()[0]);
+                        if (row['CN'] == '12') { if (row ['species'] == 'Mg')  { A_avg_Mg = r_ion * r; }}
```

Comment:
reduce code duplication by using a dictionary for `A_avg`

Reply:
done